File: Rust/src/year2019/intcode_computer/opcode.rs
use std::convert::From;
use core::fmt;
use std::fmt::Error;

#[derive(PartialEq, Eq)]
pub enum Opcode
{
    Add = 1,
    Multiply = 2,
    Input = 3,
    Output = 4,
    JumpIfTrue = 5,
    JumpIfFalse = 6,
    LessThan = 7,
    Equals = 8,
    AdjustRelativeBase = 9,
    Halt = 99,
}

impl fmt::Display for Opcode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result
    {
        match *self {
            Opcode::Add                => write!(f, "ADD"),
            Opcode::Multiply           => write!(f, "MUL"),
            Opcode::Input              => write!(f, "IPT"),
            Opcode::Output             => write!(f, "OPT"),
            Opcode::JumpIfTrue         => write!(f, "JIT"),
            Opcode::JumpIfFalse        => write!(f, "JIF"),
            Opcode::LessThan           => write!(f, "LTN"),
            Opcode::Equals             => write!(f, "UQA"),
            Opcode::AdjustRelativeBase => write!(f, "ARB"),
            Opcode::Halt               => write!(f, "HLT"),
            _ => panic!("Failed to format opcode.")
        }
    }
}


impl Opcode
{
    pub fn get_argument_count(&self) -> usize
    {
        match self
        {
            Opcode::Add             => return 3,
            Opcode::Multiply        => return 3,
            Opcode::Halt            => return 0,
            _ => panic!("Failed to get argument count for {}", self)
        }
    }
}


impl From<i64> for Opcode
{
    fn from(num: i64) -> Self
    {
        match num
        {
            1  => return Opcode::Add,
            2  => return Opcode::Multiply,
            3  => return Opcode::Input,
            4  => return Opcode::Output,
            5  => return Opcode::JumpIfTrue,
            6  => return Opcode::JumpIfFalse,
            7  => return Opcode::LessThan,
            8  => return Opcode::Equals,
            9  => return Opcode::AdjustRelativeBase,
            99 => return Opcode::Halt,

            _ => panic!("Failed to parse instruction {}", num)
        }
    }
}

File: C#/src/Years/Year2019/IntCodeComputer/Opcode.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Years.Year2019.IntCodeComputer
{
    public enum Opcode
    {
        Add = 1,
        Multiply = 2,
        Input = 3,
        Output = 4,
        JumpIfTrue = 5,
        JumpIfFalse = 6,
        LessThan = 7,
        Equals = 8,
        AdjustRelativeBase = 9,
        Halt = 99,
    }
    

    public static partial class Extensions
    {
        //Lookup tables for opcodes:
        //(not using dictionaries - switch cases resolve to lookup tables according to documentation)

        /// <summary>
        /// Opcode to string, for disassembly/debug. GetString to disambiguiate between .net object ToString 
        /// </summary>
        /// <param name="opcode"></param>
        /// <returns></returns>
        public static string GetString(this Opcode opcode)
        {
            switch (opcode)
            {
                case Opcode.Add               : return "ADD";
                case Opcode.Multiply          : return "MUL";
                case Opcode.Input             : return "INP";
                case Opcode.Output            : return "OUT";
                case Opcode.JumpIfTrue        : return "JIT";
                case Opcode.JumpIfFalse       : return "JIF";
                case Opcode.LessThan          : return "LTN";
                case Opcode.Equals            : return "EQL";
                case Opcode.AdjustRelativeBase: return "ARB";
                case Opcode.Halt              : return "HLT";
                default: throw new Exception("Unsupported opcode.");
            }
        }


        /// <summary>
        /// Returns the amount of arguments expected with this opcode
        /// </summary>
        /// <param name="opcode"></param>
        /// <returns></returns>
        public static int GetArgumentCount(this Opcode opcode)
        {
            switch (opcode)
            {
                case Opcode.Add               : return 3;
                case Opcode.Multiply          : return 3;
                case Opcode.Input             : return 1;
                case Opcode.Output            : return 1;
                case Opcode.JumpIfTrue        : return 2;
                case Opcode.JumpIfFalse       : return 2;
                case Opcode.LessThan          : return 3;
                case Opcode.Equals            : return 3;
                case Opcode.AdjustRelativeBase: return 1;
                case Opcode.Halt              : return 0;
                default: throw new Exception("Unsupported opcode.");
            }
        }


        /// <summary>
        /// Parses a number into an opcode
        /// </summary>
        /// <param name="number"></param>
        /// <returns></returns>
        public static Opcode ToOpcode(this long number)
        {
            switch (number)
            {
                case 1 : return Opcode.Add                  ;
                case 2 : return Opcode.Multiply             ;
                case 3 : return Opcode.Input                ;
                case 4 : return Opcode.Output               ;
                case 5 : return Opcode.JumpIfTrue           ;
                case 6 : return Opcode.JumpIfFalse          ;
                case 7 : return Opcode.LessThan             ;
                case 8 : return Opcode.Equals               ;
                case 9 : return Opcode.AdjustRelativeBase   ;
                case 99: return Opcode.Halt                 ;
                default: throw new Exception("Unsupported opcode.");
            }
        }
    }
}
